### ISSUE TITLE
Quantize Mask-RCNN

### ISSUE THREAD

**Issue by kchechil** [openvinotoolkit/nncf#20]
Quantize Mask-RCNN to INT8 so that it has <1% acc drop comparing to FP32.
This includes generation of the following models in ONNX format as output:
* model with FakeQuantize
* model with QuantizeLinear/DequantizeLinear

Accuracy results are needed as well.

cc @AlexKoff88

**Comment by vanyalzr** [comment]
Results for a MaskRCNN-ResNet50-FPN model (bbox AP on COCO):

Original model (fp32):

Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.379 
Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.592 
Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.411

Mixed mode quantized (activations asymmetric, weights symmetric):

 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.376
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.589
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.409

Fully symmetric (both activations and weights) quantization:

Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.372
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.583
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.406

**Comment by Wang-zipeng** [comment]
> Quantize Mask-RCNN to INT8 so that it has <1% acc drop comparing to FP32.
> This includes generation of the following models in ONNX format as output:
> 
> * model with FakeQuantize
> * model with QuantizeLinear/DequantizeLinear
> 
> Accuracy results are needed as well.
> 
> cc @AlexKoff88

We are trying to do the same thing as you. Would you please share where did you change in config/code.  Because official implement not include Instance Segmentation yet. Thank you.

**Comment by AlexKoff88** [comment]
@vanyalzr, can you please help with this?

**Comment by vanyalzr** [comment]
Hi @Wang-zipeng, 

You can use NNCF on top of mmdetection with instance segmentation models available there. We provide a patch for mmdetection with NNCF integrated into it in `third_party_integration/mmdetection`. I've made a PR with an updated patch that contains an example config for an INT8 quantized Mask-RCNN-R50-FPN model here: #174 

Let us know if there is any other specific instance segmentation model from mmdetection that you wish to compress and we'll provide the config.

**Comment by Wang-zipeng** [comment]
@vanyalzr Thank you for your reply. I have been try your commit, it works fluency on single-GPU machine, but it have two small problems on multi-GPU machine and my solutions are as follows:
/mmdet/apis/train.py:
from mmcv.runner import (HOOKS, DistSamplerSeedHook, EpochBasedRunner,
                         build_optimizer, load_checkpoint, OptimizerHook)
configs/mask_rcnn/mask_rcnn_r50_fpn_1x_coco_int8.py
add "find_unused_parameters=True"
Training looks normal so far, and 2nd epoch result are as follows:
Accumulating evaluation results...
DONE (t=9.72s).
Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.154
Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=1000 ] = 0.323
Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=1000 ] = 0.129
Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=1000 ] = 0.083
Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=1000 ] = 0.183
Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=1000 ] = 0.198
Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.300
Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=300 ] = 0.300
Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=1000 ] = 0.300
Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=1000 ] = 0.141
Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=1000 ] = 0.331
Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=1000 ] = 0.405
2020-10-09 13:33:38,019 - mmdet - INFO - Evaluating segm...
Loading and preparing results...
DONE (t=2.02s)
creating index...
index created!
Running per image evaluation...
Evaluate annotation type *segm*
DONE (t=59.56s).
Accumulating evaluation results...
DONE (t=9.72s).
Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.144
Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=1000 ] = 0.290
Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=1000 ] = 0.127
Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=1000 ] = 0.064
Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=1000 ] = 0.168
Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=1000 ] = 0.199
Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.268
Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=300 ] = 0.268
Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=1000 ] = 0.268
Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=1000 ] = 0.111
Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=1000 ] = 0.295
Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=1000 ] = 0.376

**Comment by Wang-zipeng** [comment]
@vanyalzr I have another question:
1. Can this model convert to IR by openvino? We want run it on cpu
2. training image size is (1333, 800) and image size in dummy_forward is (800, 800), Is this conflict?

**Comment by Wang-zipeng** [comment]
Hi @vanyalzr，
I am trying your config to train mask_rcnn, but I can't reach your performance as your statement, and as the epoch increase performance decrease. My machine is 8-GPU, Is this the core?

**Comment by wdkwyf** [comment]
hi, @Wang-zipeng and @vanyalzr  , my epoch-2 bbx AP is:
Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.256
Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=1000 ] = 0.448
Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=1000 ] = 0.263

which is higher than yours:

> Average Precision (AP) @[ IoU=0.50:0.95 | area= all | maxDets=100 ] = 0.144
> Average Precision (AP) @[ IoU=0.50 | area= all | maxDets=1000 ] = 0.290
> Average Precision (AP) @[ IoU=0.75 | area= all | maxDets=1000 ] = 0.127

**Comment by Wang-zipeng** [comment]
@wdkwyf so you can reimplement result?

**Comment by Wang-zipeng** [comment]
@Wang-zipeng There is something wrong in my setting, i forgot to set the pretrain model to coco pretrain, and if i set the coco pretrain model, my 2nd epoch result is 
Average Precision (AP) @[ IoU=0.50:0.95 | area= all | maxDets=100 ] = 0.238
Average Precision (AP) @[ IoU=0.50 | area= all | maxDets=1000 ] = 0.392
Average Precision (AP) @[ IoU=0.75 | area= all | maxDets=1000 ] = 0.253
lower than yours, maybe is my training machine is 8-GPU, Do you have question that as epoch increase, performance decrease.
My result is :
1st epoch "bbox_mAP": 0.265, 2nd epoch "bbox_mAP": 0.238, 3rd "bbox_mAP": 0.213, 4th "bbox_mAP": 0.197

**Comment by wdkwyf** [comment]
hi, @Wang-zipeng . I only tried two epoch for now:
AP for IoU=0.50:0.95  IoU=0.50   IoU=0.75 is:
1st epoch:  0.207 0.391 0.201
2nd epoch: 0.256 0.448 0.263

**Comment by Wang-zipeng** [comment]
@wdkwyf 
How is your training, i can get a bbox result on single-GPU machine as follows:
1st epoch: 0.208 0.388 0.204
2nd epoch: 0.247 0.44 0.252
3rd epoch: 0.268 0.459 0.282
4th epoch: 0.27 0.459 0.284
5th epoch: 0.283 0.48 0.3
I guess your result almost the same and also can't get the result 0.37~

**Comment by vanyalzr** [comment]
Hi @Wang-zipeng @wdkwyf,

I think I found the root cause of such low mAP values in your case - the config I provided originally was for the *pytorch-style* models in mmdetection, and you probably grabbed the best checkpoint to start [from here](https://github.com/open-mmlab/mmdetection/tree/master/configs/mask_rcnn) (the one with 40.8 bbox mAP), which is for a *caffe-style* model. I was able to reproduce values similar to yours with the pytorch-style config and a caffe-style checkpoint (degradation to ~0.20-0.25 mAP).

I've updated #174 so that you can start from this caffe-style checkpoint and actually got pretty good results: 

metrics right after quantization (before fine-tuning): 
bbox Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.386
segm Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.354

after 1 epoch of fine-tuning:
bbox Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.4020
segm Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.3640

You can see that a drop of less 0.01 can be reached for both bbox and segm mAPs after 1 epoch of tuning.

**Comment by Wang-zipeng** [comment]
@vanyalzr 
Thank you for your reply. I think i tried pytorch-style config and pytorch-style pretrain model. Do you mean i must use caffe-style config and caffe-style model can achieve that result, and pytorch-style config and model can't got that result? Or both of them(caffe-style and pytorch-style) can achieve pretty good results? I will try in a few days.

**Comment by Wang-zipeng** [comment]
@vanyalzr 
I think i can got a good result by caffe-style config and model, and epoch 1 result are as follows:
bbox:
Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.403
Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=1000 ] = 0.608
Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=1000 ] = 0.440
Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=1000 ] = 0.237
Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=1000 ] = 0.436
Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=1000 ] = 0.527
Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.538
Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=300 ] = 0.538
Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=1000 ] = 0.538
Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=1000 ] = 0.357
Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=1000 ] = 0.573
Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=1000 ] = 0.670
segm:
Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.365
Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=1000 ] = 0.578
Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=1000 ] = 0.389
Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=1000 ] = 0.199
Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=1000 ] = 0.394
Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=1000 ] = 0.497
Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.488
Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=300 ] = 0.488
Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=1000 ] = 0.488
Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=1000 ] = 0.306
Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=1000 ] = 0.520
Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=1000 ] = 0.631

**Comment by kchechil** [comment]
@vanyalzr could we close this?

**Comment by vanyalzr** [comment]
I believe since the accuracy issue has been resolved, we should also address the other questions raised above.
Concerning the input image size - this does not matter that much for dummy forward, since it is only used to trace the model graph in NNCF. The input size to be used should be specified upon ONNX export or, alternatively, the dynamic axes should be used to allow flexible input size.
@wdkwyf @Wang-zipeng I opened another issue #285 to make sure that Mask-RCNN export to ONNX and then to OpenVINO IR works properly.
